### PR TITLE
feat: toolbar updates for price impacts

### PR DESCRIPTION
**Pull request by just-toby**
- use a new Caption type for price impacts
- update colors for price impact messages
- fix bug with displaying price impact below the output field

image

image

image

image

### CODE REVIEW

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/Swap/Toolbar/index.tsx [hunk 1]
```diff
@@ -71,8 +71,10 @@ export default memo(function Toolbar() {
         return <Caption.InsufficientLiquidity />
       }
       if (trade?.inputAmount && trade.outputAmount) {
-        return (
-          <Caption.Trade trade={trade} outputUSDC={outputUSDC} impact={impact} gasUseEstimateUSD={gasUseEstimateUSD} />
+        return impact ? (
+          <Caption.PriceImpactWarning impact={impact} trade={trade} />
```
[the remark this answers is not in the document]
Comment (just-toby) [reply src/components/Swap/Toolbar/index.tsx:75]: thanks, good catch. i think i can rely on `impact.warning` here since it's doing those checks.

File: src/components/Swap/Toolbar/Caption.tsx [hunk 2]
```diff
@@ -183,32 +179,44 @@ export function WrapCurrency({ inputCurrency, outputCurrency, gasUseEstimateUSD
...
         tooltip={{
+          content: <RoutingDiagram trade={trade} />,
+        }}
+      />
+      <GasEstimate gasUseEstimateUSD={gasUseEstimateUSD} />
+    </>
+  )
+}
+
+type PriceImpactWarningProps = Omit<TradeProps, 'gasUseEstimateUSD'> & { impact: PriceImpact }
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:199]: nit: It'd also be appropriate here to combine types everywhere, eg:

    export interface TradeProps {...}
    export function Trade({...}: TradeProps & GasEstimateProps) {
    export interface PriceImpactProps extends TradeProps {...}

I'd prefer this to using Omit, because it's less complex overall. wdyt?

File: src/components/Swap/Toolbar/Caption.tsx [hunk 3]
```diff
@@ -183,32 +179,44 @@ export function WrapCurrency({ inputCurrency, outputCurrency, gasUseEstimateUSD
...
+
+export function PriceImpactWarning({ impact }: PriceImpactWarningProps) {
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:201]: I'd rather this just be called `PriceImpact`, considering the warning is checked in Toolbar/index.tsx, not here.

File: src/components/Swap/Toolbar/Caption.tsx [hunk 4]
```diff
@@ -62,9 +60,7 @@ function Caption({ icon: Icon = AlertTriangle, caption, color = 'secondary', too
 
 function GasEstimate({ gasUseEstimateUSD }: GasEstimateProps) {
   const isWideWidget = useIsWideWidget()
-  const displayEstimate = !gasUseEstimateUSD
-    ? '-'
-    : formatCurrencyAmount({ amount: gasUseEstimateUSD, isUsdPrice: true })
+  const displayEstimate = formatCurrencyAmount(gasUseEstimateUSD, NumberType.FiatGasPrice)
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:63]: This feels like a dangerous change to make in the same PR. Are you able to move it to its own PR, once conedison lands?
What does formatCurrencyAmount return if !gasUseEstimateUSD?

**Review by zzmp — CHANGES_REQUESTED** [review]
The only real blockers are verifying the impact color in the Output section, and potentially separating the GasEstimate change. Otherwise just minor nits - looks great!

File: src/components/Swap/Toolbar/Caption.tsx [hunk 5]
```diff
@@ -196,9 +204,11 @@ export function Trade({ trade, outputUSDC, gasUseEstimateUSD }: TradeProps) {
   )
 }
 
-type PriceImpactWarningProps = Omit<TradeProps, 'gasUseEstimateUSD'> & { impact: PriceImpact }
+interface PriceImpactProps {
+  impact: PriceImpactValue
+}
 
-export function PriceImpactWarning({ impact }: PriceImpactWarningProps) {
+export function PriceImpact({ impact }: TradeProps & PriceImpactProps) {
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:211]: You don't need TradeProps here at all. You don't need to back out the changes to WrapCurrencyProps or TradeProps, though.

nit: WrapCurrency could just be called Wrap (similarly, WrapCurrencyProps->WrapProps).

File: src/components/Swap/Toolbar/Caption.tsx [hunk 6]
```diff
@@ -1,12 +1,11 @@
...
 import { Placement } from '@popperjs/core'
+import { formatPriceImpact } from '@uniswap/conedison/format'
 import { Currency, CurrencyAmount, Token } from '@uniswap/sdk-core'
-import Column from 'components/Column'
 import Row from 'components/Row'
-import Rule from 'components/Rule'
 import Tooltip from 'components/Tooltip'
 import { loadingCss } from 'css/loading'
-import { PriceImpact } from 'hooks/usePriceImpact'
+import { PriceImpact as PriceImpactValue } from 'hooks/usePriceImpact'
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:8]: This isn't a concrete value, it's a type. What about `IPriceImpact` (`I` for Interface), or `PriceImpactType`?

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/Swap/Toolbar/Caption.tsx (hunk 6 above)
Comment (just-toby) [reply src/components/Swap/Toolbar/Caption.tsx:8]: thanks. names...

**Review by tinaszheng — APPROVED** [review]
v nice
